File: vdie/src/calculators/eta.js
'use strict';

/**
 * ETA Calculator — Haversine distance + Speed Over Ground
 *
 * Usage:
 *   const { calculateEta } = require('./eta');
 *   const result = calculateEta(fromLat, fromLon, toLat, toLon, sogKnots);
 */

const EARTH_RADIUS_NM = 3440.065; // nautical miles

/**
 * Haversine formula — returns great-circle distance in nautical miles.
 * @param {number} lat1
 * @param {number} lon1
 * @param {number} lat2
 * @param {number} lon2
 * @returns {number} distance in nautical miles
 */
function haversineNm(lat1, lon1, lat2, lon2) {
  const toRad = (deg) => (deg * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return EARTH_RADIUS_NM * c;
}

/**
 * Calculate ETA from current position to destination point.
 *
 * @param {number} fromLat - current latitude
 * @param {number} fromLon - current longitude
 * @param {number} toLat   - destination latitude
 * @param {number} toLon   - destination longitude
 * @param {number} sogKnots - speed over ground in knots
 * @returns {{
 *   distanceNm: number,
 *   etaHours: number,
 *   calculatedEtaUtc: string|null,
 *   etaConfidence: 'HIGH'|'MEDIUM'|'LOW'|'UNAVAILABLE'
 * }}
 */
function calculateEta(fromLat, fromLon, toLat, toLon, sogKnots) {
  if (
    toLat == null ||
    toLon == null ||
    fromLat == null ||
    fromLon == null
  ) {
    return { distanceNm: null, etaHours: null, calculatedEtaUtc: null, etaConfidence: 'UNAVAILABLE' };
  }

  const distanceNm = haversineNm(fromLat, fromLon, toLat, toLon);

  // Cannot calculate ETA if vessel is not moving
  if (!sogKnots || sogKnots < 0.5) {
    return { distanceNm, etaHours: null, calculatedEtaUtc: null, etaConfidence: 'LOW' };
  }

  const etaHours = distanceNm / sogKnots;
  const etaMs = etaHours * 3600 * 1000;
  const calculatedEtaUtc = new Date(Date.now() + etaMs).toISOString();

  let etaConfidence;
  if (sogKnots >= 5 && distanceNm < 5000) {
    etaConfidence = 'HIGH';
  } else if (sogKnots >= 1) {
    etaConfidence = 'MEDIUM';
  } else {
    etaConfidence = 'LOW';
  }

  return { distanceNm, etaHours, calculatedEtaUtc, etaConfidence };
}

module.exports = { haversineNm, calculateEta };

File: pfe/src/eta.py
import logging
import os
from datetime import datetime, timedelta, timezone

LOG = logging.getLogger("pfe")

ETA_ROUTING_FACTOR = float(os.getenv("ETA_ROUTING_FACTOR", 1.08))


def haversine_nm(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    from math import atan2, cos, radians, sin, sqrt

    R = 3440.065
    dlat = radians(lat2 - lat1)
    dlon = radians(lon2 - lon1)
    a = (
        sin(dlat / 2) ** 2
        + cos(radians(lat1)) * cos(radians(lat2)) * sin(dlon / 2) ** 2
    )
    return R * 2 * atan2(sqrt(a), sqrt(1 - a))


def calculate_eta(waypoints: list[dict], sog_knots: float, vessel_id: str) -> dict:
    total_nm = sum(
        haversine_nm(
            waypoints[i]["lat"],
            waypoints[i]["lon"],
            waypoints[i + 1]["lat"],
            waypoints[i + 1]["lon"],
        )
        for i in range(len(waypoints) - 1)
    )

    if sog_knots < 0.5:
        LOG.warning(
            {
                "event": "ETA_UNAVAILABLE",
                "vesselId": vessel_id,
                "sogKnots": sog_knots,
                "reason": "Vessel not underway",
            }
        )
        return {
            "distanceNm": round(total_nm, 1),
            "etaUtc": None,
            "etaHours": None,
            "etaConfidence": "UNAVAILABLE",
        }

    adjusted_nm = total_nm * ETA_ROUTING_FACTOR
    eta_hours = adjusted_nm / sog_knots
    eta_dt = datetime.now(timezone.utc) + timedelta(hours=eta_hours)

    LOG.info(
        {
            "event": "ETA_CALCULATED",
            "vesselId": vessel_id,
            "distanceNm": round(total_nm, 1),
            "sogKnots": sog_knots,
            "etaHours": round(eta_hours, 2),
            "etaUtc": eta_dt.isoformat(),
            "routingFactor": ETA_ROUTING_FACTOR,
        }
    )

    return {
        "distanceNm": round(total_nm, 1),
        "etaUtc": eta_dt.isoformat(),
        "etaHours": round(eta_hours, 2),
        "etaConfidence": "HIGH" if sog_knots > 2.0 else "MEDIUM",
    }
